lib/rethinkdb/connection/multiplexer.ex

```diff
@@ -0,0 +1,214 @@
+  """
+  use GenServer
+
+  import DBConnection.ConnectionError, only: [exception: 1]
+
+  @doc false
+  defstruct conn: {nil, nil}, token: 1, ets: nil, ref: nil
+
+  @doc """
+  Starts a `Connection` process linked to the current process.
+  """
+  def start_link(options \\ []) do
+    GenServer.start_link(__MODULE__, options, [])
+  end
+
+  @doc """
+  Sends the given `message` to the server.
+
+  Options:
+
+  * `:timeout` - maximum time to wait for a reply (default: `5_000`).
+  * `:noreply` - do not wait for the server acknowledgement of the message.
+  """
+  def send_recv(pid, message, options \\ []) do
+    noreply = Keyword.get(options, :noreply, false)
+    timeout = Keyword.get(options, :timeout, 5_000)
+    if noreply do
+      GenServer.cast(pid, {:send, message})
+    else
+      token = Keyword.get(options, :token)
+      {:ok, GenServer.call(pid, {:send, token, message}, timeout)}
+    end
+  end
+
+  #
+  # GenServer API
+  #
+
+  def init(options) do
+    host = Keyword.get(options, :hostname, "localhost")
+    port = Keyword.get(options, :port, 28015)
+    user = Keyword.get(options, :user, "admin")
+    pass = Keyword.get(options, :password, "")
+
+    {transport, sock_opts} =
+      if ssl_opts = Keyword.get(options, :ssl) do
+        {:ssl, Enum.map(Keyword.fetch!(ssl_opts, :ca_certs),  &({:cacertfile, &1})) ++ [verify: :verify_peer]}
+      else
+        {:gen_tcp, []}
+      end
+
+    # Connects to the server and perform a handshake
+    sock_opts = [packet: :raw, mode: :binary, active: :false] ++ sock_opts
+    with {:ok, sock} <- transport.connect(String.to_char_list(host), port, sock_opts),
+         {:ok, conn} <- init_connection(transport, sock),
+         {:ok, _nil} <- handshake(conn, user, pass),
+         {_pid, ref} <- spawn_monitor(__MODULE__, :recv_loop, [self, conn]),
+    do:  {:ok, %__MODULE__{conn: conn, ets: :ets.new(__MODULE__, [:set, :private]), ref: ref}}
+  end
+
+  def handle_call({:send, nil, msg}, from, %__MODULE__{conn: conn, token: token, ets: table} = state) do
+    :ok = send_payload(conn, token, msg)
+    :ets.insert(table, {token, from})
+    {:noreply, %__MODULE__{state | token: token + 1}}
+  end
+
+  def handle_call({:send, token, msg}, from, %__MODULE__{conn: conn, ets: table} = state) do
+    :ok = send_payload(conn, token, msg)
+    :ets.insert(table, {token, from})
+    {:noreply, state}
+  end
+
+  def handle_cast({:send, msg}, %__MODULE__{conn: conn, token: token} = state) do
+    :ok = send_payload(conn, token, msg)
+    {:noreply, %__MODULE__{state | token: token + 1}}
+  end
+
+  def handle_cast({:reply, token, msg}, %__MODULE__{ets: table} = state) do
+    :ets.take(table, token)
+    |> List.first
+    |> elem(1)
+    |> GenServer.reply({token, msg})
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, _ref, :process, _pid, reason}, state) do
+    {:stop, reason, state}
+  end
+
+  @doc false
+  def recv_loop(pid, {transport, sock} = conn) do
```

Comment:
Take a look at my original code https://github.com/hamiltop/rethinkdb-elixir/blob/master/lib/rethinkdb/connection/request.ex#L24 . It's a bit complex because you don't how much data you will be receiving so you have to buffer in the process state.